rustc: use 1.82.0 binary to bootstrap rustc 1.82.0

skeuchel opened the pull request:
This uses the upstream version 1.82.0 binary to bootstrap the rust compiler of the same version instead of the binary of the previous version on riscv64-linux. The reason is that the older 1.81.0 cannot bootstrap the new version for this platform as discussed in https://github.com/rust-lang/rust/issues/129268. This is a temporary fix that is only necessary for this version and it only applies to native compilation, not to cross compilation. 

Edit: Update it for all platforms.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] riscv64-linux
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alyssais commented on pkgs/development/compilers/rust/1_82.nix:
```diff
@@ -105,7 +105,13 @@ import ./default.nix
+      # 1.81.0 binary cannot build the 1.82.0 compiler on that platform.
+      # https://github.com/rust-lang/rust/issues/129268.
+      # Remove this when bumping to 1.83.0.
+      if stdenv.targetPlatform.isRiscV64 then "1.82.0" else "1.81.0";
```
Shouldn't this be buildPlatform?  Whether the 1.82.0 tarball is used depends on whether we're building _from_ RISC-V, right?

But also, maybe it would be better to just use 1.82.0 on all platforms for this release, to avoid confusing divergences like this?  I'm pretty sure the comment above that says "MUST be one version prior" is not accurate, because my understanding is that rustc also always supports building a released compiler with itself.

skeuchel commented:
> But also, maybe it would be better to just use 1.82.0 on all platforms for this release, to avoid confusing divergences like this? I'm pretty sure the comment above that says "MUST be one version prior" is not accurate, because my understanding is that rustc also always supports building a released compiler with itself.

Maintainers what do you think?

alyssais commented:
FWIW while I'm not listed as a maintainer, that's basically an accident. I'm generally the one who handles rustc upgrades and a lot of other rust stuff.

[Apparently](https://github.com/NixOS/nixpkgs/commit/ba9c71b999b1aa28dd2af6d23359d2db7b5da8ac) in 2017 we were told that we should be using the previous version of rustc, not the current one. @workingjubilee do you know if that's still what we should be preferring in general?

workingjubilee commented:
> But also, maybe it would be better to just use 1.82.0 on all platforms for this release, to avoid confusing divergences like this? I'm pretty sure the comment above that says "MUST be one version prior" is not accurate, because my understanding is that rustc also always supports building a released compiler with itself.

Correct, the Rust compiler must be able to build itself.

workingjubilee commented:
In 2017, our usual flow for bootstrapping the compiler was different, so it could have been much rockier, and it has always been touchy on 32-bit.

Nowadays, for development purposes, we build the standard library with the last beta version. From release to release, that means "the last version". This was easier to standardize on, and have everyone reuse it. Then we build the compiler with the beta + new library. Then it builds its own library again to complete its sysroot.

Then, after building the compiler with the beta, there is another build with the new compiler. In many cases there are no practical differences.

skeuchel committed:
rustc: use 1.82.0 binary to bootstrap rustc 1.82.0

Change the bootstrap on all platforms to use a version of the upstream binary
that is equal to the version that is being built, instead of a binary of one
version prior. This is necessary for the 1.82.0 release because on some
platforms (https://github.com/rust-lang/rust/issues/129268) the 1.81.0 binary
can not build the 1.82.0 source. Furthermore, upstream reported that their
development process (now) usually involves building a release compiler with
itself and that "the Rust compiler must be able to build itself".

skeuchel commented:
So if upstream now ensures that each version can build itself, then I propose to change the bootstrap accordingly. I have updated the PR to do so.

alyssais commented:
Could you put an explanation of why we're making this change in the commit message please?

winterqt commented on pkgs/development/compilers/rust/1_82.nix:
```diff
@@ -103,25 +103,23 @@ import ./default.nix
       else
         llvmPackages_18;
 
-    # Note: the version MUST be one version prior to the version we're
-    # building
```
It may be worth writing a similar comment here, e.g. "the version must be the same version that we're building".

skeuchel replied:
Good point. Done.

winterqt commented:
Thank you!
